initial check in

weiyuan-jiang opened the pull request:
## Types of change(s)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Trivial change (affects only documentation or cleanup)
- [ ] Refactor (no functional changes, no api changes)

## Checklist
- [ ] Tested this change with a run of GEOSgcm
- [ ] Ran the Unit Tests (`make tests`)

## Description
Add EASEGridFactory so the Regridder can use EASEGrid
## Related Issue

mathomp4 commented:
@weiyuan-jiang It looks like there is some ambiguity in some GEOS code:
```
[ 83%] Building Fortran object src/Components/@GEOSgcm_GridComp/GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Utils/Raster/makebcs/CMakeFiles/mkEASETilesParam.x.dir/mkEASETilesParam.F90.o
/root/project/GEOSgcm/src/Components/@GEOSgcm_GridComp/GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Utils/Raster/makebcs/mkCatchParam.F90:193:23:

  193 |       call ease_extent(Gridname, nc_ease, nr_ease )
      |                       1
Error: Name 'ease_extent' at (1) is an ambiguous reference to 'ease_extent' from module 'mapl_easeconvmod'
make[2]: *** [src/Components/@GEOSgcm_GridComp/GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Utils/Raster/makebcs/CMakeFiles/mkCatchParam.x.dir/build.make:75: src/Components/@GEOSgcm_GridComp/GEOSagcm_GridComp/GEOSphysics_GridComp/GEOSsurface_GridComp/Utils/Raster/makebcs/CMakeFiles/mkCatchParam.x.dir/mkCatchParam.F90.o] Error 1
```

Will this MAPL PR need to go with a GEOSgcm_GridComp PR?

tclune commented on base/MAPL_EASEConv.F90:
```diff
@@ -0,0 +1,818 @@
+  
+  ! =====================================================================================
+  !
+  ! EASE_conv.F90 - FORTRAN routines for conversion of Equal-Area Scalable Earth (EASE)
```
```suggestion
  ! EASE_conv.F90 - Fortran routines for conversion of Equal-Area Scalable Earth (EASE)
```

tclune commented on base/MAPL_EASEConv.F90:
```diff
@@ -0,0 +1,818 @@
+  ! CHANGELOG (easeV1_conv.F90):
+  ! ============================
+  !
+  ! easeV1_conv.F90 - FORTRAN routines for conversion of azimuthal 
```
```suggestion
  ! easeV1_conv.F90 - Fortran routines for conversion of azimuthal 
```

tclune commented on base/MAPL_EASEConv.F90:
```diff
@@ -0,0 +1,818 @@
+    
+  ! radius of the earth (km), authalic sphere based on International datum 
+  
+  real*8, parameter :: easeV1_RE_km                    = 6371.228
```
`real*4` and `real*8` not permitted in MAPL.

```suggestion
  real(kind=REAL64), parameter :: easeV1_RE_km                    = 6371.228
```

tclune commented on base/MAPL_EASEConv.F90:
```diff
@@ -0,0 +1,818 @@
+
+module MAPL_EASEConvMod
```
We have a convention for module names and file names.  Be sure to follow it.  no `Mod`

I also suggest expanding the abbreviation `Conv`: electrons are cheap.
```suggestion
module MAPL_EASEConversion
```

tclune commented on base/MAPL_EASEConv.F90:
```diff
@@ -0,0 +1,818 @@
+    else if (index(EASELabel,'M01') /=0 ) then
+       grid='M01'
+    else
+       print*,"ease_convert(): unknown grid projection and resolution: "//trim(EASELabel)//"  STOPPING."
+       stop
```
Use error handling.

tclune commented on base/MAPL_EASEConv.F90:
```diff
@@ -0,0 +1,818 @@
+    integer :: cols, rows
+    real*8  :: Rg, phi, lam, rho, CELL_km, r0, s0
+    
+    real*8, parameter :: PI = easeV1_PI
```
How many values of pi does mapl need?

weiyuan-jiang replied:
I have changed to MAPL's PI and all the REAL64  parameters' definition.  However,  I believe the change will make it non-zero diff, especially in BCS. I may have to change it back to the 'wrong' way for zero-diff.

tclune replied:
For the parameters, we have found that with optimization on, there usually is no difference.    The compiler apparently sees the advantage of not truncating just to promote.  But ... no guarantees.

For PI - if there is a difference that would be bad.   I hope not.

If this is a problem for @gmao-rreichle, we can relax this requirement for this PR and just add it to the list of things to fix in MAPL3.

tclune replied:
@gmao-rreichle For PI - yes I don't really have an argument other than it is what it is.   But for the other parameters, I suspect you are getting different values with debug vs non-debug.   I've not looked closely though.  If all of the literals are real integers (e.g., 123.0) then it may not matter.    But ... in the past, MAPL had things like:
```
real(kind=REAL64), parameter :: pi = 3.14159265...
```
And the value depended on optimization until we fixed it as:
```
real(kind=REAL64), parameter :: pi = 3.14159265...d+0
```

tclune commented on base/MAPL_EASEConv.F90:
```diff
@@ -0,0 +1,818 @@
+       r = r0 + Rg * lam * easeV1_COS_PHI1
+       s = s0 - Rg * sin(phi) / easeV1_COS_PHI1
+       
+    endif
```
Missing default clause?

weiyuan-jiang commented on base/MAPL_EASEConv.F90:
```diff
@@ -0,0 +1,818 @@
+  real*8, parameter :: M_map_reference_latitude        =   0.0
+  real*8, parameter :: M_map_second_reference_latitude =  30.0
+  
+  real*8, parameter :: M_sin_phi1 = sin(M_map_second_reference_latitude*easeV2_PI/180.)
```
I may make lead to non-zero diff

weiyuan-jiang commented:
@gmao-rreichle , I have moved the file EASE_Conv.F90 from makebcs to here ( MAPL_EASEConversion.F90). I also changed the parameters to a real double ( ie. add d0 at the end of a number)  and MAPL's PI is also used. But it may lead to non-zero-diff in BCS. Do you think it is worthy to make such changes? I can change back to the original 'wrong' way.

https://github.com/GEOS-ESM/MAPL/blob/60098729d5ab75fdf93ae0eb501b6487602a3fa6/base/MAPL_EASEConversion.F90#L101-L141

gmao-rreichle commented on base/MAPL_EASEConversion.F90:
```diff
@@ -0,0 +1,828 @@
+  !
+  ! ==========================================================================
+  use, intrinsic :: iso_fortran_env, only: REAL64
+  use MAPL_Constants, only: PI => MAPL_PI_r8
```
We have to keep the v1- and v2-specific PI values.  Otherwise, we will lose backward compatibility for creating bcs, and we would also have lat/lons in published SMAP L4 products that aren't consistent with SMAP L1-3 products published by other institutions.

gmao-rreichle commented on base/MAPL_EASEConversion.F90:
```diff
@@ -0,0 +1,828 @@
+    ! 
+    ! recall that EASE grid indexing is zero-based
+
+    if (present(ll_lat))  call ease_inverse(EASElabel, 0., rows-0.5, ll_lat, tmplon, _RC)     
+    if (present(ur_lat))  call ease_inverse(EASElabel, cols-1.0,     -0.5, ur_lat, tmplon, _RC)     
```
These two calls **only** get the **latitude** values.  I think we should stick with using 0. for the longitude index in both cases, as in the original code from EASE_conv.F90:
```
    if (present(ll_lat))  call ease_inverse(EASElabel, 0., rows-0.5, ll_lat, tmplon)     
    if (present(ur_lat))  call ease_inverse(EASElabel, 0.,     -0.5, ur_lat, tmplon)     
```
The **lon** values are set in the next two lines.

gmao-rreichle commented on base/MAPL_EASEConversion.F90:
```diff
@@ -0,0 +1,828 @@
+    if (present(ll_lat))  call ease_inverse(EASElabel, 0., rows-0.5, ll_lat, tmplon, _RC)     
+    if (present(ur_lat))  call ease_inverse(EASElabel, cols-1.0,     -0.5, ur_lat, tmplon, _RC)     
+    
+    if (present(ll_lon))  ll_lon = -180.
+    if (present(ur_lon))  ur_lon =  180.
```
The ur and ll values set here are for the "**global**" cylindrical EASE grid.  (I'm putting "global" into parentheses because the EASE min/max lat is near 85S/N.)  That is, the subroutine ease_extent does not work for "regional" grids.  There is some code related to "regional" grids in MAPL_EASEGridFactory.F90.  Is this cut-and-paste left over from the lat/lon grid that is not meant to work, or is the "regional" grid meant to work for the cylindrical EASE grids?  If it's meant to work, are we sure that subroutine ease_extent() is ok as is?

gmao-rreichle replied:
Thanks for the explanation.  If the "global" cylindrical EASE grid is considered a "regional" grid for MAPL purposes, we need to add this info as a comment to MAPL_EASEGridFactory.F90.  There should then also be checks on whether variable "regional" is indeed "true".  E.g., in MAPL_EASEGridFactory.F90 near line 340, there's the following chunk of code 
```
      regional  = (dateline == 'XY')
      if (regional) then
         delta = (this%lon_range%max - this%lon_range%min) / this%im_world
         min_coord = this%lon_range%min + delta/2
         max_coord = this%lon_range%max - delta/2
      else
         delta = 360.d0 / this%im_world
         min_coord = -180.d0 + delta/2
         max_coord = +180.d0 - delta/2
      end if
``` 
Given the fact that the EASE grid must be regional, the "else" statement should exit with an error.  There are probably other places that need the same cleanup.

gmao-rreichle commented on base/MAPL_EASEGridFactory.F90:
```diff
@@ -0,0 +1,1866 @@
+      end if
+
+      ! Cannot assume that lats and lons are evenly spaced
+      this%is_regular = .false.
```
"is_regular" may be misleading for the EASE grid.  In my understanding, the EASE grid is a "regular" grid.  It does not have even lat spacing, but it's still a regular grid (as opposed to, say, the cube-sphere tile space).    Maybe "is_evenspaced" (or something to that effect) would be a less confusing variable name.

gmao-rreichle commented on base/MAPL_EASEConversion.F90:
```diff
@@ -0,0 +1,828 @@
+
+  end subroutine easeV2_get_params
+
+  function ease_grid_name(cols, rc) result(name)
```
The function "ease_grid_name" is new and was not in the original EASE_conv.F90 module. I have a couple of concerns re. this new function:
1. The original design calls for knowing the grid name, and then subroutine "ease_extent" can be used to determine the number of columns and rows.  If we know the number of cols, we must already know the grid name (assuming we have dealt with the legacy "SMAP-EASEv[x]-M[xx]" grid name -- see my other comment).  So it doesn't make sense to me that we even need a function that converts the number of cols into the grid name. 
2. Hard-coding the relationship between the number of cols and the grid name here duplicates what's already hard-coded into ease_extent.  We shouldn't hard-code these things in two different places.
3. "ease_grid_name" requires that each EASE grid has a unique number of cols, which is ok with the handful of grids defined here but can't be generally assumed.

gmao-rreichle commented on base/NCIO.F90:
```diff
@@ -5569,6 +5571,11 @@ subroutine MAPL_ReadTilingASCII(layout, FileName, GridName, NT, im, jm, n_Grids,
           allocate(AVR(NT,9), STAT=STATUS) ! 9 columns for EASE grid
           _VERIFY(STATUS)
           allocate(AVR_transpose(9,NT))
+          ! WY note: In the old tile file, 
+          ! EASE grid name is in the form of SMAP-EASEvx-Mxx
+          ! It is changed to a standard form as EASEvx_Mxx
+          Correct_ease_name = ease_grid_name(IM(1))
+          GridNAME(1) = Correct_ease_name
```
I'm concerned about going through the number of cols to identify the grid.  See my comment near Line ~788 of MAPL_EASEConversion.F90.  Here, if we encounter the old naming convention ("SMAP-EASE..."),  it would be simple to directly convert the string SMAP-EASEv[x]-M[xx] into EASEv[x]_M[xx].

gmao-rreichle commented on base/MAPL_EASEGridFactory.F90:
```diff
@@ -0,0 +1,1866 @@
+            end if
+         end if
+
+         grid_name = ease_grid_name(im)
```
I'm not sure why we need the EASE grid name here, and how it could be obtained without going through the number of cols.  Perhaps this is unavoidable, but I still like to think that we can obtain the name without going through the number of columns.  Same for the other "initialize_from...()" subroutine below.

weiyuan-jiang replied:
At some point, if only cols is provided, we can construct the factory and thus EASE grid

gmao-rreichle reviewed (commented):
@weiyuan-jiang, @tclune : Thanks for the PR!  To answer your question about PI -- we absolutely can **not** change the original PI values, sorry.  See below for additional inline comments.

tclune commented:
> @weiyuan-jiang, @tclune : Thanks for the PR! To answer your question about PI -- we absolutely can **not** change the original PI values, sorry. See below for additional inline comments.

@gmao-rreichle  We should at least consider the forward path - are you locked into "wrong" values permanently?  E.g., what if we had to stop using Intel compiler in favor of say NVIDIA in the future?

weiyuan-jiang commented on base/MAPL_EASEGridFactory.F90:
```diff
@@ -0,0 +1,1866 @@
+            max_coord = 360.d0 - delta
+         case ('GE')
+            min_coord = delta/2
+            max_coord = 360.d0 - delta/2
```
cleaned

gmao-rreichle commented on base/MAPL_EASEGridFactory.F90:
```diff
@@ -0,0 +1,1866 @@
+! Document PE, PC, DC, DE, GC
+
+! This module generates ESMF_Grids corresponding to _regular_ lat-lon coordinate grids.
+! I.e., spacing between lats (lons) is constant.
```
It still says: "This module generates ESMF_Grids corresponding to _regular_ lat-lon coordinate grids. I.e., spacing between lats (lons) is constant."  This makes no sense.

gmao-rreichle commented on base/MAPL_EASEConversion.F90:
```diff
@@ -0,0 +1,828 @@
+#include "MAPL_ErrLog.h"
+
+module MAPL_EASEConversion
+  
+  ! =====================================================================================
+  !
+  ! EASE_conv.F90 - Fortran routines for conversion of Equal-Area Scalable Earth (EASE)
```
I don't see such a change.

weiyuan-jiang commented on base/MAPL_EASEGridFactory.F90:
```diff
@@ -0,0 +1,1845 @@
+            this%lon_corners(im+1) = (this%lon_centers(im) + this%lon_centers(1))/2 + 180
+         end if
+
+         ! This section about pole/dateline is probably not needed in file data case.
+         if (abs(this%lon_corners(1) + 180) < 1000*epsilon(1.0)) then
+            this%dateline = 'DE'
+         else ! assume 'XY'
+            this%dateline = 'XY' 
+            this%lon_range = RealMinMax(this%lon_centers(1), this%lon_centers(jm))
+         end if
```
I leave 'XY' there intentionally in case in the future we need regional longitudes

weiyuan-jiang commented on base/MAPL_EASEGridFactory.F90:
```diff
@@ -0,0 +1,1845 @@
+            end if
+         end if
+
+        ! TODO: if 'lat' and 'lon' are not present then 
+        ! assume ... pole/dateline are ?
+
+        ! TODO: check radians vs degrees.  Assume degrees for now.
+
+
+        ! TODO: modify CoordinateVariable so that get_coordinate_data() is overloaded
+        ! for different types (as subroutine) to avoid casting here.
+        ! TODO: add get_coordinate_variable() interface to avoid the need to cast
```
NO. We assume if it is initialized from metadata,  the data we need is there.  I removed those "TODO" comments

weiyuan-jiang commented on base/MAPL_EASEGridFactory.F90:
```diff
@@ -0,0 +1,1845 @@
+
+         ! Corners are the midpoints of centers (and extrapolated at the 
+         ! poles for lats.)
+         if (.not. allocated(this%lon_corners)) then 
```
indent corrected

biljanaorescanin commented:
@gmao-rreichle PR was tested for BCS package creation for `v12 EASE M09` and `v13 EASE M36` and it is zero diff. 
To run make_bcs needs 3 branches: 
`MAPL              | (t) v2.56.0 (DH)           | (b) feature/wjiang/ease_grid`
`GEOSldas_GridComp | (b) develop (DH, 4c7de26)  | (b) feature/wjiang/ease_grid`
`GEOSgcm_GridComp  | (b) develop (DH, 85b060ab) | (b) feature/wjiang/ease_grid`

biljanaorescanin commented:
> Thanks, @biljanaorescanin, for the test results. Did you run make_bcs from a build of GEOSgcm or GEOSldas?

@gmao-rreichle  I've tested from GEOSldas. Do you need me to do the same from GEOSgcm?

weiyuan-jiang commented:
> > Thanks, @biljanaorescanin, for the test results. Did you run make_bcs from a build of GEOSgcm or GEOSldas?
> 
> @gmao-rreichle I've tested from GEOSldas. Do you need me to do the same from GEOSgcm?

Do you test bcs generated by these PRs? That is important

biljanaorescanin commented:
> Do you test bcs generated by these PRs? That is important

Yes, few comments up I say I used those 3 branches for bcs creation. 
I can share experiment outputs on chat if you need to see it. 

This was my testing clone: 
https://github.com/GEOS-ESM/GEOSldas/blob/testing/borescan_ease_factory/components.yaml

gmao-rreichle commented:
> > > Thanks, @biljanaorescanin, for the test results. Did you run make_bcs from a build of GEOSgcm or GEOSldas?
> > 
> > 
> > @gmao-rreichle I've tested from GEOSldas. Do you need me to do the same from GEOSgcm?

No, make_bcs should be identical when run from either App.  I mostly wanted to know if you worked within GEOSldas, and how the GEOSldas_GridComp branch fits into all of this.  This has now been addressed with new PRs for GEOSldas and GEOSldas_GridComp.  I added a new test requirement to the intro comment above (0-diff for GEOSldas).

gmao-rreichle reviewed (approved):
@mathomp4, @weiyuan-jiang : The new MAPL EASE Grid Factory has been successfully  0-diff tested with GEOSldas and GEOSgcm.  The PR is ready to be merged as far as I'm concerned.  
To use the new functionality in GEOSldas, we'll need a new MAPL release.  Both GEOSldas and GEOSgcm are on MAPL v2.56, so the new MAPL release should work ok for both. 
I'm a bit unsure about where we are with GEOSadas (which now includes the LDAS in its build).  I know Sara was sorting out MAPL versions for ADAS and LDAS recently, so whatever the issue was then might still be an issue now.  But we can sort this out later. 
cc: @tclune @biljanaorescanin
